UniStuttgart-VISUS/megamol
# Frontend 1: ImagePresentation Service manages Graph Entry Points

Refactoring of PR #755. Depends on PR #775.
The start of this PR is commit 9c27a01 "add ImagePresentation Service: pokes rendering of graph entry points"

* Reorganize management of Graph Entry Points in Frontend around ImagePresentation Service - collects rendering results from views in textures
* Proof of Concept: View3DGL and View2DGL output rendering result as `ImageWrapper` object to ImagePresentation Service
* Depends on PR #748
* Prepares frontend for PR #744 

## Summary of Changes
Pull MegaMolGraph Entry Points into separate Service - the Image Presentation -, give each Graph Entry Point a Texture that the corresponding View writes rendering results into.
This enables rendering and screenshots independent of GLFW Window size and will be implemented in a later PR. In the long term, enables sophisticated Split-View like layouting of View results in the ImGui or using some other layouting mechanism. 

## References and Context
Depends on refactoring of all relevant Views to export their FBO. The Frontend<->View interface minimizes data copying by simply wrapping the texture object of the view using a `ImageWrapper` object, which takes a non-owning reference to the wrapped texture. Currently, only View3DGL and View2DGL are hooked into the frontend image mechanism, but connecting more views should not be too difficult (see `GetRenderingResult()` in `AbstractView.h`, `View3DGL.h` and `View2DGL.h`).

The frontend images are designed as zero-copy wrappers (`ImageWrapper`) and are able to generalize over Views that render images into CPU Buffers and into OpenGL Textures. 

## Test Instructions
Not immediately testable since this PR provides infrastructure for the Camera redesign. Make sure that all your important projects load and render correctly.
